Implement `extract_volume_patches` (issue #17843)

hsgkim opened the pull request:
Implement `extract_volume_patches` (issue #17843). `rates` not supported as of now.

The code is pretty similar to that of `extract_image_patches`. I'm not quite sure as to whether adding the `pbtxt` file alone suffices for the documentation; I would appreciate it if someone could comment on this.

hsgkim committed:
Implement extract_volume_patches

hsgkim commented:
Sorry, I forgot to tag relevant people:
@tatianashp @asimshankar @ezhulenev
Could you have a look at this please? Thanks in advance!

tatianashp commented:
@hsgkim Thank you for this PR.

ezhulenev commented:
@hsgkim I’m with limited access to internet right now, will take a look at this PR on monday, but from my phone everything seems ok. Though I have one question, why did you commented all the code related to rates? If it works and it’s an api consistent with image patches, why not to enable it right now?

hsgkim commented:
@ezhulenev As mentioned before in the issue, none of Eigen's `extract_volume_patches` functions let one pass `rates` as an argument; it's all fixed to `(1, 1, 1, 1, 1)`. I left them as comments in case someone finds a workaround for that and decides to add `rates`.

ezhulenev commented on tensorflow/core/kernels/extract_volume_patches_op.cc:
```diff
@@ -0,0 +1,189 @@
+    const int stride_cols = strides_[3];
+
+    /*
+    // In order to enable rates, uncomment the following lines and use
```
Can you add a brief note about missing rates parameter in Eigen op, as a reason for commenting this code. Maybe even a TODO, adding rates to Eigen should be a reasonably small change.

ezhulenev commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -2549,6 +2549,109 @@ REGISTER_OP("ExtractImagePatches")
+      }
+
+      /*
+      std::vector<int32> rates;
```
+a note here, maybe just a reference to the op.cc

ezhulenev commented:
Already forgot about that Eigen thing :) adding some comments should help people like me

hsgkim commented:
Will do, thanks for the feedback :D!

hsgkim committed:
comments regarding why rates are disabled

hsgkim commented:
Added comments.

hsgkim commented:
@ezhulenev @akshaym I'm really sorry for bothering you, but it says feedback/copybara's check failed. But the Details hyperlink is just linked to `http://cl/210001543` and I can't figure out what went wrong. I can build tf with my commits and the tests pass; can I ignore copybara or do I have to fix something?

ezhulenev commented:
`tensorflow/tools/api/tests:api_compatibility_test` is failing (as expected)

see https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/tools/api/tests/README.txt for required steps

hsgkim committed:
update golden & pylint

hsgkim commented:
@ezhulenev Thanks! Updated golden and I disabled the pylint check just in case; it kept complaining about me adding spaces for test result `np.array`s (hinders readability otherwise).

hsgkim committed:
change golden/v2/tensorflow.pbtxt

Running the API compatibility test only checks for pbtxt files under directory
v1. Manually added extract_volume_patches under v2 as extract_image_patches is
registered under v2 as well.

hsgkim commented:
@ezhulenev Sorry for bothering you, but can I get a progress update on this please? It says the check has been pending for a week and I was wondering if it's something I did wrong. Thanks!

ezhulenev commented:
@hsgkim It's stuck in internal copybara import, whenever it gets through, I'll approve it and it should get merged (it will need additional review from tf-api owners, but that should be fast for this PR)

ezhulenev commented:
@hsgkim I remember about this PR, will find out tomorrow why it's still not available internally for review

ezhulenev commented on tensorflow/core/api_def/base_api/api_def_ExtractVolumePatches.pbtxt:
```diff
@@ -0,0 +1,49 @@
+op {
+  graph_op_name: "ExtractVolumePatches"
+  in_arg {
+    name: "images"
```
`images` makes too strong assumption about the type of the input, just `input` is better, or maybe you have any other name specific for volume/cuboid convolutions?

ezhulenev commented on tensorflow/core/api_def/base_api/api_def_ExtractVolumePatches.pbtxt:
```diff
@@ -0,0 +1,49 @@
+  in_arg {
+    name: "images"
+    description: <<END
+5-D Tensor with shape `[batch, in_rows, in_cols, depth]`.
```
You missed `in_planes` here.

ezhulenev commented on tensorflow/core/api_def/base_api/api_def_ExtractVolumePatches.pbtxt:
```diff
@@ -0,0 +1,49 @@
+    description: <<END
+5-D Tensor with shape `[batch, out_planes, out_rows, out_cols, 
+ksize_planes * ksize_rows * ksize_cols * depth]` containing image 
+patches with size `ksize_patches x ksize_rows x ksize_cols x depth` 
```
`ksize_patches` -> `ksize_planes` ?

ezhulenev reviewed (changes requested):
Sorry for another round of review, spotted it during our internal review process. This time it should be faster.

hsgkim committed:
fix documentation errors

hsgkim commented:
@ezhulenev I fixed all of them; sorry for such silly mistakes :(

ezhulenev commented on tensorflow/core/ops/array_ops.cc:
```diff
@@ -2549,6 +2549,112 @@ REGISTER_OP("ExtractImagePatches")
+// as the second parameter of all GetWindowedOutputSizeVerbose calls instead
+// of ksize_*.
+REGISTER_OP("ExtractVolumePatches")
+    .Input("images: T")
```
`input`

ezhulenev commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -1060,6 +1060,10 @@ tf_module {
     name: "extract_image_patches"
     argspec: "args=[\'images\', \'ksizes\', \'strides\', \'rates\', \'padding\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
```
`input`

ezhulenev commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -1060,6 +1060,10 @@ tf_module {
   }
+  member_method {
+    name: "extract_volume_patches"
+    argspec: "args=[\'images\', \'ksizes\', \'strides\', \'padding\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
```
`input`

ezhulenev commented on tensorflow/tools/api/golden/v2/tensorflow.pbtxt:
```diff
@@ -1060,6 +1060,10 @@ tf_module {
     name: "extract_image_patches"
     argspec: "args=[\'images\', \'ksizes\', \'strides\', \'rates\', \'padding\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
```
`input`

ezhulenev commented on tensorflow/tools/api/golden/v2/tensorflow.pbtxt:
```diff
@@ -1060,6 +1060,10 @@ tf_module {
   }
+  member_method {
+    name: "extract_volume_patches"
+    argspec: "args=[\'images\', \'ksizes\', \'strides\', \'padding\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
```
`input`

ezhulenev reviewed (changes requested):
Missing `images` -> `input`

hsgkim committed:
fix argument name

hsgkim commented:
@ezhulenev Fixed `array_ops.cc` and `golden`.

hsgkim commented:
@ezhulenev I changed the name of the input (from `images` to `input`) for everything but the test Python file; I felt iffy about using `input` as a name of an argument as it already is a name of a Python function. I checked the committed files and it looked like I didn't have anything else to change, what do you think? Thanks again in advance :)

hsgkim commented:
@martinwicke Could you do an API review please? Thanks!

martinwicke commented:
@ezhulenev can you check why the internal submit is stuck? Looks like a lint problem maybe.

martinwicke commented:
I'm triggering tests to check whether we can see the problem here.

hsgkim committed:
clang-format

hsgkim commented:
Fixed lint issues.
[GPU Python3](https://source.cloud.google.com/results/invocations/fb3ee2ec-bcd6-4777-8fda-a1adaacee97c/targets) and [XLA](https://source.cloud.google.com/results/invocations/bf32cdde-1d08-400d-86dc-3baf0bd7f2bb/targets) fails don't look like they were caused by me?

hsgkim commented:
Is the feedback/copybara check different when it is run by kokoro-team? feedback/copybara passed for commit b3ec2ca (fix argument name) before but it says it failed after kokoro-team ran it again.

martinwicke commented:
Yes this can be merged.

hsgkim commented:
@ezhulenev It looks like the test is stuck again, do I have to fix something else? Sorry for the bother.

ezhulenev commented:
@martinwicke do you know what's happening? cr/211828838 seems approved, any way to force submit?

martinwicke commented:
The tool got stuck. Sometimes, removing and adding an approval can help. It's unstuck now and submitting, should come through soon.
